crates/matrix-sdk/src/test_utils/mocks.rs

```diff
@@ -0,0 +1,514 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Helpers to mock a server and have a client automatically connected to that
+//! server, for the purpose of integration tests.
+
+#![allow(missing_debug_implementations)]
+
+use std::sync::{Arc, Mutex};
+
+use matrix_sdk_base::deserialized_responses::TimelineEvent;
+use matrix_sdk_test::{
+    test_json, InvitedRoomBuilder, JoinedRoomBuilder, KnockedRoomBuilder, LeftRoomBuilder,
+    SyncResponseBuilder,
+};
+use ruma::{MxcUri, OwnedEventId, OwnedRoomId, RoomId};
+use serde_json::json;
+use wiremock::{
+    matchers::{body_partial_json, header, method, path, path_regex},
+    Mock, MockBuilder, MockGuard, MockServer, Respond, ResponseTemplate, Times,
+};
+
+use super::logged_in_client;
+use crate::{Client, Room};
+
+/// A `wiremock` server along a client connected it, with useful methods to help
+/// mocking Matrix client-server API endpoints easily.
+///
+/// This is a pair of a MockServer and a Client (one can retrieve them
+/// respectively with [`Self::server()`] and [`Self::client()`]).
+///
+/// It implements mock endpoints, expanding and limited the shared code as much
+/// as possible, so the mocks are still flexible to use as scoped/unscoped
+/// mounts, named, and so on.
+///
+///It works like this:
+///
+/// - start with telling which endpoint you'd like to mock, e.g.
```

Comment:
```suggestion
/// - start by saying which endpoint you'd like to mock, e.g.
```

["telling" is transitive: you have to say *who* you are telling about a thing.]